subcommands/targets/add.go

```diff
@@ -0,0 +1,288 @@
+)
+
+// Consider moving the `Target` type along with its receivers to the `client` package
+// or do opposite - "hide" it
+type (
+	Target struct {
+		Length int64             `json:"length"`
+		Hashes tuf.Hashes        `json:"hashes"`
+		Custom *client.TufCustom `json:"custom,omitempty"`
+	}
+	Targets map[string]*Target
+)
+
+func NewTarget(tufMeta tuf.FileMeta) (*Target, error) {
+	custom, err := api.TargetCustom(tufMeta)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal Target custom from TUF meta: %s", err.Error())
+	}
+	return &Target{
+		Length: tufMeta.Length,
+		Hashes: tufMeta.Hashes,
+		Custom: custom,
+	}, nil
+}
+
+func (t *Target) Derive(newVer int) *Target {
+	derivedCustom := *t.Custom
+	derivedCustom.Version = strconv.Itoa(newVer)
+	derivedCustom.CreatedAt = time.Now().UTC().Format(time.RFC3339)
+	derivedCustom.UpdatedAt = derivedCustom.CreatedAt
+	return &Target{
+		Length: t.Length,
+		Hashes: t.Hashes,
+		Custom: &derivedCustom,
+	}
+}
+
+func (t *Target) HardwareId() string {
+	return t.Custom.HardwareIds[0]
+}
+
+func (t *Target) Version() int {
+	ver, err := strconv.Atoi(t.Custom.Version)
+	if err != nil {
+		panic(fmt.Sprintf("failed to convert Target version to int: %s", err.Error()))
+	}
+	return ver
+}
+
+func (t *Target) Tags() []string {
+	return t.Custom.Tags
+}
+
+func (t *Target) HasTag(tag string) bool {
+	for _, ct := range t.Tags() {
+		if ct == tag {
+			return true
+		}
+	}
+	return false
+}
+
+func (t *Target) Name() string {
+	return t.Custom.Name + "-" + t.Custom.Version
+}
+
+func (t *Target) SetHash(hash string) error {
+	hb, err := base64.StdEncoding.DecodeString(hash)
+	if err != nil {
+		return err
+	}
+	t.Hashes[notary.SHA256] = hb
+	return nil
+}
+
+func init() {
+	addCmd := &cobra.Command{
+		Use:   "add",
+		Short: "Add target(s)",
+		Run:   doAdd,
+		Long:  "",
+	}
+	cmd.AddCommand(addCmd)
+	addCmd.Flags().StringVarP(&addTargetType, "type", "", "", "Target type")
+	addCmd.Flags().StringVarP(&addTags, "tags", "", "", "comma,separate,list of Target tags")
+	addCmd.Flags().StringVarP(&addSrcTag, "src-tag", "", "", "OSTree Target tag to base app targets on")
+	addCmd.Flags().StringVarP(&addOutFile, "out", "o", "", "a file path to output the new Targets to")
+	addCmd.Flags().BoolVarP(&addPrint, "print", "", false, "print generated new Targets to stdout")
+	addCmd.Flags().BoolVarP(&addDryRun, "dry-run", "", false, "don't post generated new Targets")
+}
+
+func doAdd(cmd *cobra.Command, args []string) {
+	factory := viper.GetString("factory")
+	supportedTargetTypes := map[string]func(factory string, tags []string, srcTag string, args []string) (Targets, error){
+		"app":    createAppTargets,
+		"ostree": createOstreeTarget,
+	}
+	if len(addTargetType) == 0 {
+		subcommands.DieNotNil(errors.New("missing mandatory flag `--type`"))
+	}
+	deriveTargets, ok := supportedTargetTypes[addTargetType]
+	if !ok {
+		subcommands.DieNotNil(errors.New("unsupported type of Target: " + addTargetType))
```

Comment:
I think this change might clean this up a little - https://github.com/foundriesio/fioctl/pull/195

Reply:
It looks quite clean to me :). Also, I am not sure if it can help much since a number of variables can be different. A user can add many Apps or many ostree hashes.

Reply:
the change is for enforcing enums in what arguments a flag will accept. so you could drop the whole if statement and let cobra handle it.

Reply:
I still need that `if !ok` since the `supportedTargetTypes` map can be incorrectly defined.

Reply:
However, it's not runtime error then, rather assert.

Reply:
I actually couldn't find how cobra can enforce enums value for a flag. Even if it was there I still need that `if !ok` just to make sure that the `supportedTargetTypes` map is correctly defined and a correct function is returned.
I looked into the way Volodya handles `--role` flag for the `key` command and it actually uses `switch case` statement to accomplish the same goal that I've achieved with the map  and the `if !ok` statement.